airflow/models/xcom_arg.py

```diff
@@ -17,714 +17,100 @@
-            yield from arg.iter_references()
-
-    def concat(self, *others: XComArg) -> ConcatXComArg:
-        # Flatten foo.concat(x).concat(y) into one call.
-        return ConcatXComArg([*self.args, *others])
-
-    def get_task_map_length(self, run_id: str, *, session: Session) -> int | None:
-        all_lengths = (arg.get_task_map_length(run_id, session=session) for arg in self.args)
-        ready_lengths = [length for length in all_lengths if length is not None]
-        if len(ready_lengths) != len(self.args):
-            return None  # If any of the referenced XComs is not ready, we are not ready either.
-        return sum(ready_lengths)
-
-    @provide_session
-    def resolve(
-        self, context: Mapping[str, Any], session: Session = NEW_SESSION, *, include_xcom: bool = True
-    ) -> Any:
-        values = [arg.resolve(context, session=session, include_xcom=include_xcom) for arg in self.args]
-        for value in values:
-            if not isinstance(value, (Sequence, dict)):
-                raise ValueError(f"XCom concat expects sequence or dict, not {type(value).__name__}")
-        return _ConcatResult(values)
+    return session.scalar(query)
 
 
-_XCOM_ARG_TYPES: Mapping[str, type[XComArg]] = {
-    "": PlainXComArg,
-    "concat": ConcatXComArg,
-    "map": MapXComArg,
-    "zip": ZipXComArg,
-}
+@get_task_map_length.register
+def _(xcom_arg: MapXComArg, run_id: str, *, session: Session):
+    return get_task_map_length(xcom_arg.arg, run_id, session=session)
 
 
-def serialize_xcom_arg(value: XComArg) -> dict[str, Any]:
-    """DAG serialization interface."""
-    key = next(k for k, v in _XCOM_ARG_TYPES.items() if isinstance(value, v))
-    if key:
-        return {"type": key, **value._serialize()}
-    return value._serialize()
+@get_task_map_length.register
+def _(xcom_arg: ZipXComArg, run_id: str, *, session: Session):
+    all_lengths = (get_task_map_length(arg, run_id, session=session) for arg in xcom_arg.args)
+    ready_lengths = [length for length in all_lengths if length is not None]
+    if len(ready_lengths) != len(xcom_arg.args):
+        return None  # If any of the referenced XComs is not ready, we are not ready either.
+    if isinstance(xcom_arg.fillvalue, ArgNotSet):
+        return min(ready_lengths)
+    return max(ready_lengths)
 
 
-def deserialize_xcom_arg(data: dict[str, Any], dag: DAG) -> XComArg:
-    """DAG serialization interface."""
-    klass = _XCOM_ARG_TYPES[data.get("type", "")]
-    return klass._deserialize(data, dag)
+@get_task_map_length.register
+def _(xcom_arg: ConcatXComArg, run_id: str, *, session: Session):
+    all_lengths = (get_task_map_length(arg, run_id, session=session) for arg in xcom_arg.args)
+    ready_lengths = [length for length in all_lengths if length is not None]
+    if len(ready_lengths) != len(xcom_arg.args):
+        return None  # If any of the referenced XComs is not ready, we are not ready either.
+    return sum(ready_lengths)
```

Comment:
```suggestion
@get_task_map_length.register
def _(xcom_arg: ConcatXComArg, run_id: str, *, session: Session):
    ready_lengths = []
    for arg in xcom_arg.args:
        length = get_task_map_length(arg, run_id, session=session)
        if length is None:
            return None
        ready_lengths.append(length)
    return sum(ready_lengths)
```

would be simpler and better, since all we are checking for is `None` 🤷 ?

Reply:
Maybe -- I didn't really look at these functions, but just moved them from the XComArg subclasses as they were, ConcatXComArg in this case.